<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <!-- We resume the logic in empty_world.launch, changing only the name of the world to be launched -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="amazonrobot_1_warehouse.world"/> <!-- Note: the world_name is with respect to GAZEBO_RESOURCE_PATH environmental variable -->
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="true"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
    <arg name="verbose" default="true"/>
  </include>

  <!-- Used to load map generated by gmapping -->
  <node name="map_server" pkg="map_server" type="map_server" args="./maps/map.yaml">
    <param name="frame_id" value="map"/>
  </node>

  <!-- Nodes for the main robot -->

  <!-- Main robot laser to base transform -->
  <node pkg="tf" type="static_transform_publisher" name="base_to_laser_main" 
    args="0.139000 0.000000 0.15 0.000000 0.000000 0.000000 /amazon_warehouse_robot/amazon_warehouse_robot::chassis /amazon_warehouse_robot/hokuyo_link 20" />

  <node pkg="tf" type="static_transform_publisher" name="map_to_odom_main" 
    args="0 0 0 0 0 0 /map /amazon_warehouse_robot/odom 20" />

  <!-- Multiplexer is used to choose one controller at a time -->
  <!-- The only publisher at /amazon_warehouse/cmd_vel -->
  <node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager"/>
  <node pkg="nodelet" type="nodelet" name="cmd_vel_mux" args="load yocs_cmd_vel_mux/CmdVelMuxNodelet nodelet_manager">
    <remap from="/cmd_vel_mux/output/cmd_vel"           to="/amazon_warehouse_robot/cmd_vel" />
    <param name="yaml_cfg_file" value="./config/mux_params.yaml"/>
  </node>

  <!-- Move base node. Used for global and local navigation -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <rosparam file="./config/main_robot/costmap_common_params.yaml" command="load" ns="global_costmap" /> 
    <rosparam file="./config/main_robot/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="./config/main_robot/local_costmap_params.yaml" command="load" />
    <rosparam file="./config/main_robot/global_costmap_params.yaml" command="load" /> 
    <rosparam file="./config/main_robot/base_local_planner_params.yaml" command="load" />

    <remap from="/odom" to="/amazon_warehouse_robot/odom"/>

    <remap from="/move_base/NavfnROS/plan"                                         to="/amazon_warehouse_robot/move_base/NavfnROS/plan" />
    <remap from="/move_base/TrajectoryPlannerROS/cost_cloud"                       to="/amazon_warehouse_robot/move_base/TrajectoryPlannerROS/cost_cloud" />
    <remap from="/move_base/TrajectoryPlannerROS/global_plan"                      to="/amazon_warehouse_robot/move_base/TrajectoryPlannerROS/global_plan" />
    <remap from="/move_base/TrajectoryPlannerROS/local_plan"                       to="/amazon_warehouse_robot/move_base/TrajectoryPlannerROS/local_plan" />
    <remap from="/move_base/TrajectoryPlannerROS/parameter_descriptions"           to="/amazon_warehouse_robot/move_base/TrajectoryPlannerROS/parameter_descriptions" />
    <remap from="/move_base/TrajectoryPlannerROS/parameter_updates"                to="/amazon_warehouse_robot/move_base/TrajectoryPlannerROS/parameter_updates" />

    <remap from="/cmd_vel"                                                         to="/amazon_warehouse_robot/move_base/cmd_vel" />
    <remap from="/move_base/cancel"                                                to="/amazon_warehouse_robot/move_base/cancel" />
    <remap from="/move_base/current_goal"                                          to="/amazon_warehouse_robot/move_base/current_goal" />
    <remap from="/move_base/feedback"                                              to="/amazon_warehouse_robot/move_base/feedback" />

    <remap from="/move_base/global_costmap/costmap"                                to="/amazon_warehouse_robot/move_base/global_costmap/costmap" />
    <remap from="/move_base/global_costmap/costmap_updates"                        to="/amazon_warehouse_robot/move_base/global_costmap/costmap_updates" />
    <remap from="/move_base/global_costmap/footprint"                              to="/amazon_warehouse_robot/move_base/global_costmap/footprint" />
    <remap from="/move_base/global_costmap/inflation_layer/parameter_descriptions" to="/amazon_warehouse_robot/move_base/global_costmap/inflation_layer/parameter_descriptions" />
    <remap from="/move_base/global_costmap/inflation_layer/parameter_updates"      to="/amazon_warehouse_robot/move_base/global_costmap/inflation_layer/parameter_updates" />
    <remap from="/move_base/global_costmap/obstacle_layer/parameter_descriptions"  to="/amazon_warehouse_robot/move_base/global_costmap/obstacle_layer/parameter_descriptions" />
    <remap from="/move_base/global_costmap/obstacle_layer/parameter_updates"       to="/amazon_warehouse_robot/move_base/global_costmap/obstacle_layer/parameter_updates" />
    <remap from="/move_base/global_costmap/parameter_descriptions"                 to="/amazon_warehouse_robot/move_base/global_costmap/parameter_descriptions" />
    <remap from="/move_base/global_costmap/parameter_updates"                      to="/amazon_warehouse_robot/move_base/global_costmap/parameter_updates" />
    <remap from="/move_base/global_costmap/static_layer/parameter_descriptions"    to="/amazon_warehouse_robot/move_base/global_costmap/static_layer/parameter_descriptions" />
    <remap from="/move_base/global_costmap/static_layer/parameter_updates"         to="/amazon_warehouse_robot/move_base/global_costmap/static_layer/parameter_updates" />

    <remap from="/move_base/goal"                                                  to="/amazon_warehouse_robot/move_base/goal" />

    <remap from="/move_base/local_costmap/costmap"                                 to="/amazon_warehouse_robot/move_base/local_costmap/costmap" />
    <remap from="/move_base/local_costmap/costmap_updates"                         to="/amazon_warehouse_robot/move_base/local_costmap/costmap_updates" />
    <remap from="/move_base/local_costmap/footprint"                               to="/amazon_warehouse_robot/move_base/local_costmap/footprint" />
    <remap from="/move_base/local_costmap/inflation_layer/parameter_descriptions"  to="/amazon_warehouse_robot/move_base/local_costmap/inflation_layer/parameter_descriptions" />
    <remap from="/move_base/local_costmap/inflation_layer/parameter_updates"       to="/amazon_warehouse_robot/move_base/local_costmap/inflation_layer/parameter_updates" />
    <remap from="/move_base/local_costmap/obstacle_layer/parameter_descriptions"   to="/amazon_warehouse_robot/move_base/local_costmap/obstacle_layer/parameter_descriptions" />
    <remap from="/move_base/local_costmap/obstacle_layer/parameter_updates"        to="/amazon_warehouse_robot/move_base/local_costmap/obstacle_layer/parameter_updates" />
    <remap from="/move_base/local_costmap/parameter_descriptions"                  to="/amazon_warehouse_robot/move_base/local_costmap/parameter_descriptions" />
    <remap from="/move_base/local_costmap/parameter_updates"                       to="/amazon_warehouse_robot/move_base/local_costmap/parameter_updates" />

    <remap from="/move_base/parameter_descriptions"                                to="/amazon_warehouse_robot/move_base/parameter_descriptions" />
    <remap from="/move_base/parameter_updates"                                     to="/amazon_warehouse_robot/move_base/parameter_updates" />
    <remap from="/move_base/result"                                                to="/amazon_warehouse_robot/move_base/result" />
    <remap from="/move_base/status"                                                to="/amazon_warehouse_robot/move_base/status" />
    <remap from="/move_base_simple/goal"                                           to="/amazon_warehouse_robot/move_base_simple/goal" />
  </node>

</launch>
